### PR TITLE
fix: :bug: fix sar saved model export

### PR DESCRIPTION
**Pull request by khalidMindee**
This PR solves #602 for SAR recognition models to allow export as a saved model.

### CODE REVIEW

File: doctr/models/recognition/sar/tensorflow.py [hunk 1]
```diff
@@ -127,16 +127,15 @@ def call(
...
         # run first step of lstm
         # holistic: shape (N, rnn_units)
         _, states = self.lstm_decoder(holistic, states, **kwargs)
         # Initialize with the index of virtual START symbol (placed after <eos> so that the one-hot is only zeros)
-        symbol = tf.fill(features.shape[0], self.vocab_size + 1)
+        symbol = tf.fill(tf.shape(features)[:1], self.vocab_size + 1)
         logits_list = []
-        if kwargs.get('training') and gt is None:
-            raise ValueError('Need to provide labels during training for teacher forcing')
+
```
Comment (fg-mindee) [inline doctr/models/recognition/sar/tensorflow.py:138]: This needs to stay, was it blocking the savedmodel export?

File: doctr/models/recognition/sar/tensorflow.py [hunk 2]
```diff
@@ -150,7 +149,8 @@ def call(
             # shape (N, rnn_units + 1) -> (N, vocab_size + 1)
             logits = self.output_dense(logits, **kwargs)
             # update symbol with predicted logits for t+1 step
-            if kwargs.get('training'):
+            if kwargs.get('training') and gt is not None:
+                #'Need to provide labels during training for teacher forcing'
```
Comment (fg-mindee) [inline doctr/models/recognition/sar/tensorflow.py:153]: no need to duplicate comment

File: doctr/models/recognition/sar/tensorflow.py [hunk 3]
```diff
@@ -150,7 +149,8 @@ def call(
             # shape (N, rnn_units + 1) -> (N, vocab_size + 1)
             logits = self.output_dense(logits, **kwargs)
             # update symbol with predicted logits for t+1 step
-            if kwargs.get('training'):
+            if kwargs.get('training') and gt is not None:
```
Comment (fg-mindee) [inline doctr/models/recognition/sar/tensorflow.py:152]: can be reverted since we already ensure that the gt is provided in training mode above

File: doctr/models/recognition/sar/tensorflow.py [hunk 4]
```diff
@@ -302,10 +302,11 @@ def __call__(
...
         decoded_strings_pred = tf.strings.split(decoded_strings_pred, "<eos>")
         decoded_strings_pred = tf.sparse.to_dense(decoded_strings_pred.to_sparse(), default_value='not valid')[:, 0]
-        word_values = [word.decode() for word in decoded_strings_pred.numpy().tolist()]
-
-        return list(zip(word_values, probs.numpy().tolist()))
 
+        return {
+            "decoded_strings_pred":decoded_strings_pred,
+            "probs":probs,
+        }
```
Comment (fg-mindee) [inline doctr/models/recognition/sar/tensorflow.py:309]: We can't change the output signature of the call method in this PR or we'll break the high-level API

**Review by fg-mindee — CHANGES_REQUESTED** [review]
Thanks for the PR! I added some comments, let's avoid breaking changes in this PR if possible :pray:

**Comment by fharper** [comment]
@khalidMindee is this PR still good? If yes, can you review FG's comments, modify in consequences so we can potentially merge it please?

**Comment by khalidMindee** [comment]
Ah great @felixdittrich92 that's fine for me to close the PR .
